### ISSUE TITLE
Does passing 'training' flag to tf.keras.Sequential do anything?

### ISSUE THREAD

**Issue by milongo** [tensorflow/tensorflow#36936]
## URL(s) with the issue:
https://www.tensorflow.org/tutorials/customization/custom_training_walkthrough#define_the_loss_and_gradient_function

Please provide a link to the documentation entry, for example:
https://www.tensorflow.org/api_docs/python/tf/keras/Sequential

https://www.tensorflow.org/api_docs/python/tf/keras/Model
## Description of issue (what needs changing):
It is unclear whether the Sequential class makes use a 'training' flag fed into it during training/inference, as the tutorial above implies. 

### Clear description
When building a custom model subclassing from tf.keras.Model, the standard signature for writing the `call` is as follows: `def call(self, inputs, training=None, mask=None):`

If my class includes submodels of the form Sequential, I am able to pass this flag forward but I'm unaware whether it's doing anything, as documentation from the class doesn't mention this flag.
Looking at the customization tutorial above, however, the flag is passed into a Sequential model that does not include layers whose behavior change during training/inference. So I don't know if that flag is doing anything.

### Correct links

Is the link to the source code correct?

### Parameters defined

Are all parameters defined and formatted correctly?

### Returns defined

Are return values defined?

### Raises listed and defined

Are the errors defined? For example,
https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/feature_column/categorical_column_with_vocabulary_file#raises

### Usage example

Is there a usage example?

See the API guide: https://www.tensorflow.org/community/contribute/docs_ref
on how to write testable usage examples.

### Request visuals, if applicable

Are there currently visuals? If not, will it clarify the content?

### Submit a pull request?

Are you planning to also submit a pull request to fix the issue? See the docs
contributor guide: https://www.tensorflow.org/community/contribute/docs,
docs API guide: https://www.tensorflow.org/community/contribute/docs_ref and the
docs style guide: https://www.tensorflow.org/community/contribute/docs_style

**Comment by pandrey-fr** [comment]
> If my class includes submodels of the form Sequential, I am able to pass this flag forward but I'm unaware whether it's doing anything

Any `training` boolean flag passed to a `Sequential` model will be passed to each and every of its layers, which will trigger training-only behaviours _if any is implemented_. Typically, one will have dropout applied conditionally on this flag, in which case passing it will indeed have an effect.
Note that using the default keras training loop (_i.e._ the `fit` method) will automatically pass this flag along (while it will be set to False in `evaluate` and `predict`).

**Comment by milongo** [comment]
Thank you for your answer. Can you please point me towards a source code file where I can see this behavior? Perhaps the docs should include this information as well since it's not very clear in my opinion.

**Comment by pandrey-fr** [comment]
You can see in the definition of `tf.keras.Sequential.call` [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/sequential.py#L269) that it expects a `training` flag, and that it will indeed be passed to any sublayer whose `call` signature includes that keyword.

As for the use of `training` itself, you can notably read about it [here](https://www.tensorflow.org/guide/keras/custom_layers_and_models#privileged_training_argument_in_the_call_method) in the Keras guide :)

**Comment by gadagashwini** [comment]
@milongo, Does the @pandrey-fr's comment help in resolving your issue?

**Comment by milongo** [comment]
@gadagashwini , @pandrey-fr's comments do answer my questions, but I still feel the documentation for `tf.keras.Sequential` should include the `training` flag somewhere in the arguments i.e. 
https://www.tensorflow.org/api_docs/python/tf/keras/Sequential#arguments_21 and make it explicit that the flag is passed to subsequent layers and made use of *if* those layers have it defined in their behavior, whether they're custom or not.

**Comment by pandrey-fr** [comment]
@milongo To be clear, the flag is passed _by default_ as part of `tf.keras.Model.fit`, `tf.keras.Model.evaluate` and `tf.keras.Model.predict`. The fact that a model's call may accept a `training` argument is documented [here](https://www.tensorflow.org/api_docs/python/tf/keras/Model) :
> If you subclass Model, you can optionally have a training argument (boolean) in call, which you can use to specify a different behavior in training and inference:

The fact that `Layer` instances may accept it (and will receive it if so) is part of their API, as indicated [here](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#__call__).

That being said, it may be worth to augment the existing documentation of `Sequential` as you suggest it.

**Comment by milongo** [comment]
@pandrey-fr what is the behavior if the model is called in a custom training loop? Such as the following:

```
    for epoch in range(epochs):
        for step, (x_batch, y_batch) in enumerate(train_dataset):

            if epoch > mlp.warm_up:
                for layer in mlp.mean.layers:
                    layer.trainable = False
                for layer in mlp.variance.layers:
                    layer.trainable = True
            with tf.GradientTape() as tape:
                output = mlp(x_batch)
                loss = loss_fn(y_batch, output)
            grads = tape.gradient(loss, mlp.trainable_weights)
            optimizer.apply_gradients(zip(grads, mlp.trainable_weights))
            mean_preds, _ = output
```
You can see that I do not bother to pass the `training` flag anywhere.
Here, `mlp` is a custom class subclassed from `tf.keras.Model`, whose's `call` function is as follows:

```
    def call(self, inputs, training=None, mask=None):

        mean_predictions = []
        variance_predictions = []
        for idx in range(self.num_models):
            mean_predictions.append(self.mean[idx](inputs))
            variance_predictions.append(self.variance[idx](inputs))
        mean_stack = tf.stack(mean_predictions)
        variance_stack = tf.stack(variance_predictions)

        return mean_stack, variance_stack
```
You can see that I do not bother to pass the `training` flag anywhere.

(For clarity, `mlp` has two sets of ensembles, one in `self.mean` and another in `self.variance`, where each ensemble is defined as a list of `tf.keras.Sequential` modules):
```
self.mean = [
            tf.keras.Sequential([
                layers.Dense(512, activation='relu', input_shape=(self.input_dim,)),
                layers.Dropout(0.5),
                layers.Dense(256, activation='relu'),
                layers.Dropout(0.5),
                layers.Dense(self.output_dim)
            ])
            for _ in range(self.num_models)
        ]
```
Interestingly enough, my models train just fine, so I'm guessing the default internal value for the flag is `training=True`, even if the headers have it as `training=None`. I.e. somewhere, that flag's value is actually set to `True`.

**Comment by pandrey-fr** [comment]
As far as I know, if you do not pass `training=True` explicitly, your model should run without dropout. This may not be prejudicial to training, but if you meant to make use of it it would probably be worth it to pass the flag within your custom training loop.

**Comment by pandrey-fr** [comment]
You are welcome; I am glad this was helpful. Regarding documentation, perhaps @gadagashwini could assign the arbitration to a tensorflower; otherwise, maybe you could submit a PR to the tensorflow-docs repository?

**Comment by pandrey-fr** [comment]
You are welcome; I am glad this was helpful. Regarding documentation, perhaps @gadagashwini could assign the arbitration to a tensorflower; otherwise, maybe you could submit a PR to the tensorflow-docs repository?

**Comment by milongo** [comment]
Ahh I think you are correct. Passing the flag explicitly drastically affected training. 
Thank you! 
I could close this now, but what about the documentation issues I mentioned?

**Comment by jvishnuvardhan** [comment]
@milongo Are you planning to create any PR to update the docs? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
